go-controller/pkg/ovn/gateway.go

```diff
@@ -249,14 +249,23 @@ func (gw *GatewayManager) GatewayInit(
 	hostSubnets []*net.IPNet,
 	l3GatewayConfig *util.L3GatewayConfig,
 	sctpSupport bool,
-	gwLRPIfAddrs, drLRPIfAddrs []*net.IPNet,
+	gwLRPJoinIPs, drLRPIfAddrs []*net.IPNet,
 	externalIPs []net.IP,
 	enableGatewayMTU bool,
 ) error {
 
 	gwLRPIPs := make([]net.IP, 0)
-	for _, gwLRPIfAddr := range gwLRPIfAddrs {
-		gwLRPIPs = append(gwLRPIPs, gwLRPIfAddr.IP)
+	for _, gwLRPJoinIP := range gwLRPJoinIPs {
+		gwLRPIPs = append(gwLRPIPs, gwLRPJoinIP.IP)
+	}
+	if gw.netInfo.TopologyType() == types.Layer2Topology {
+		// At layer2 GR LRP acts as the layer3 ovn_cluster_router so we need
+		// to configure here the .1 address, this will work only for IC with
+		// one node per zone, since ARPs for .1 will not go beyond local switch.
+		// This is being done to add the ICMP SNATs for .1 podSubnet that OVN GR generates
+		for _, subnet := range hostSubnets {
+			gwLRPIPs = append(gwLRPIPs, util.GetNodeGatewayIfAddr(subnet).IP)
+		}
```

Comment:
I might be missing something, but what is the purpose of this ? 

AFAIU, you don't use the `gwLRPIPs` anywhere - well, other than to get the MAC addr of the router port in line 405, but you can use the joinIPs instead (in fact, I think it would be preferable to do that).

I think you only need (it's already being done in line 414) to compute the `gwLRPNetworks`, which, AFAIU, you're setting the GW addr into the LRP.

Reply:
the gwLRPIPs is used to compute the routes and SNATs below. The sphegetti exists in the older code... I am trying not to change much in mine

Reply:
routes:
```
for _, gwLRPIP := range gwLRPIPs {
		lrsr := nbdb.LogicalRouterStaticRoute{
			IPPrefix: gwLRPIP.String(),
			Nexthop:  gwLRPIP.String(),
		}
```
SNATs:
```
	joinNATs := make([]*nbdb.NAT, 0, len(gwLRPIPs))
	for _, gwLRPIP := range gwLRPIPs {
		externalIP, err := util.MatchIPFamily(utilnet.IsIPv6(gwLRPIP), externalIPs)
		if err != nil {
			return fmt.Errorf("failed to find valid external IP family match for join subnet IP: %s on "+
				"gateway router: %s", gwLRPIP, gatewayRouter)
		}
		joinIPNet, err := util.GetIPNetFullMask(gwLRPIP.String())
		if err != nil {
			return fmt.Errorf("failed to parse full CIDR mask for join subnet IP: %s", gwLRPIP)
		}
		nat := libovsdbops.BuildSNAT(&externalIP[0], joinIPNet, "", extIDs)
		joinNATs = append(joinNATs, nat)
	}
```

Reply:
for l2 both of these don't make much sense but the SNAT is a hack in OVNK to get around an OVN bug which got fixed but we need another PR to remove that... :( 
confusing I know.

Reply:
and on 634 we want to use the join IP or the GW IP ? https://github.com/ovn-kubernetes/ovn-kubernetes/blob/dca22499b961a2ee0f03e524e43e82e767b79f7c/go-controller/pkg/ovn/gateway.go#L634

I assume it's the join IP, but now, it could be easier to convey that info - use the joinIPs variable. Same in 737.

Indeed. Damned spaghetti.

Reply:
@maiqueb that's the hard problem here :) as of now for always it was using the gwLRPIP which included the podsubnet and joinsubnet so I cannot change that in this PR unless I know older code was wrong.